python/dataset.cpp

```diff
@@ -123,18 +166,45 @@ void init_dataset(py::module &m) {
            py::arg("data") = std::map<std::string, Variable>{},
            py::arg("coords") = std::map<Dim, Variable>{},
            py::arg("labels") = std::map<std::string, Variable>{})
+      .def(py::init([](const DatasetProxy &other) { return Dataset{other}; }))
       .def("__setitem__", [](Dataset &self, const std::string &name,
                              Variable data) { self.setData(name, data); })
       .def("__setitem__",
            [](Dataset &self, const std::string &name,
               const DataConstProxy &data) { self.setData(name, data); })
+      // TODO: nvaytet: I do not understand why this is not covered by the
+      // py::implicitly_convertible<VariableProxy, Variable>();
+      // statement in variable.cpp, but this is needed if a VariableProxy is
+      // used instead of a Variable.
+      // Maybe it's because it's in a different file?
+      .def("__setitem__",
+           [](Dataset &self, const std::string &name,
+              const VariableProxy &data) { self.setData(name, data); })
+      .def("__setitem__",
+           [](Dataset &self, const std::tuple<Dim, scipp::index> &index,
+              DatasetProxy &other) {
+             auto[dim, i] = index;
+             for (const auto[name, item] : self.slice(Slice(dim, i)))
+               item.assign(other[name]);
+           })
       .def("__setitem__",
            [](Dataset &self, const std::string &name, const DataArray &data) {
              self.setData(name, data);
            })
       .def("set_sparse_coord", &Dataset::setSparseCoord)
       .def("set_sparse_labels", &Dataset::setSparseLabels)
       .def("set_coord", &Dataset::setCoord)
+      .def("set_coord",
```

Comment:
I recommend we do not do this since we are trying to move to a more python-style API. #397 will remove `set_coord` in favor of `__setitem__`, so we can do `d.coords[Dim.X] = var`. This avoids having to remember various setter names.

Reply:
:+1: ok i'll remove the syntax from the notebooks and the tests